### PR TITLE
Components: Add VisuallyHidden component

### PR DESCRIPTION
**Pull request by mkaz**
## Description

Adds a new component for rendering ScreenReaderText and the necessary CSS.
Per @youknowriad <a href="https://github.com/WordPress/gutenberg/pull/17997#issuecomment-543584382">comment here</a>

In testing this, I noticed one of the issues also for Storybook is they weren't setup to load the style.scss along with the components. This PR also adds the necessary CSS loaders in the Storybook webpack.config.js so a story can do this to include the CSS.

```
import ScreenReaderText from '../';
import '../style.scss';

export default { title: 'ScreenReaderText', component: ScreenReaderText };

export const _default = () => (
	<>
		<ScreenReaderText>
			This should not show.
		</ScreenReaderText>
		<div>
			This text will always show.
		</div>
	</>
);
```
This is my first new component being added, so not sure all the pieces that need to be properly updated.

## How has this been tested?

- Add `<ScreenReaderText> Test text. </ScreenReaderText>` and confirm it includes a div and proper style rules.

- Run Storybook: `npm run designs-system:dev` confirm ScreenReaderText does/doesn't show

## Types of changes
- Adds new component `ScreenReaderText`
- Updates Storybook to support loading .scss

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by gziolo** [comment]
It's often named `VisuallyHidden` in UI components libraries. See:
- https://reakit.io/docs/visually-hidden/ (@diegohaz)
- https://reacttraining.com/reach-ui/visually-hidden

File: packages/components/src/screen-reader-text/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# ScreenReaderText
+
+A component used to render text intended for a screen reader, not visible elsewhere.
```
Comment (gziolo) [inline packages/components/src/screen-reader-text/README.md]: You can reference WordPress docs here, maybe also copy some parts of that document:
https://make.wordpress.org/accessibility/handbook/markup/the-css-class-screen-reader-text/

File: packages/components/src/screen-reader-text/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+### Usage
+
+```jsx
+<ScreenReaderText> Show text for screenreader. </ScreenReaderText>
```
Comment (gziolo) [inline packages/components/src/screen-reader-text/README.md]: In other places we add full examples:

```jsx
import { Dashicon } from '@wordpress/components';

const MyDashicon = () => (
	<div>
		<Dashicon icon="admin-home" />
		<Dashicon icon="products" />
		<Dashicon icon="wordpress" />
	</div>
);
```

File: packages/components/src/screen-reader-text/index.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+
+function ScreenReaderText( props ) {
+	return (
+		<div className="screen-reader-text">
```
Comment (gziolo) [inline packages/components/src/screen-reader-text/index.js]: You often will want to use a different tag name as well.

Comment (mkaz) [reply packages/components/src/screen-reader-text/index.js]: Instead of a div? I'm open to suggestions, should it b a prop?

Comment (youknowriad) [reply packages/components/src/screen-reader-text/index.js]: The className here doesn't follow our guidelines. I guess it was done this way to match the existing className. I think it's fine to change it for consistency.

Comment (youknowriad) [reply packages/components/src/screen-reader-text/index.js]: In another component we used `isInline` prop to switch between `span` and `div` to ensure HTML stays valid in all situations. I wouldn't mind starting with div and adding that prop if we find a place where we need something inline.

Comment (gziolo) [reply packages/components/src/screen-reader-text/index.js]: The thing is that WP core uses `screen-reader-text`.

See also https://github.com/WordPress/gutenberg/pull/18022#issuecomment-543974141 about `VisualHidden` proposal.

Comment (mkaz) [reply packages/components/src/screen-reader-text/index.js]: One of the reasons for using `screen-reader-text` className was to introduce the CSS into the components built-style CSS. There are several components already using that className relying on core's definition of it.

This was discovered when creating the stories which only include the build-style CSS.

Comment (youknowriad) [reply packages/components/src/screen-reader-text/index.js]: I think we should update all the usage of this class in our codebase to use this component instead, that way it's more portable, the className becomes an implementation detail. It could be done part of this PR or separately.

Comment (mkaz) [reply packages/components/src/screen-reader-text/index.js]: ok, I switched the className to `components-visualyl-hidden` 

I'll switch the other components over in a separate PR. I like keeping each PR as focused as possible and thus one adding a component doesn't really change the other behavior. Just enables the ability to fix them.

**Review by youknowriad — APPROVED** [review]

File: packages/components/src/visually-hidden/style.scss [hunk 4]
```diff
@@ -0,0 +1,13 @@
+.components-visually-hidden {
```
Comment (gziolo) [inline packages/components/src/visually-hidden/style.scss]: This is missing:
```css
.screen-reader-text:focus {
  background-color: #eee;
  clip: auto !important;
  clip-path: none;
  color: #444;
  display: block;
  font-size: 1em;
  height: auto;
  left: 5px;
  line-height: normal;
  padding: 15px 23px 14px;
  text-decoration: none;
  top: 5px;
  width: auto;
  z-index: 100000; /* Above WP toolbar. */
}
```

See https://make.wordpress.org/accessibility/handbook/markup/the-css-class-screen-reader-text/

File: packages/components/src/visually-hidden/index.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+
+function VisuallyHidden( {
+	children,
+	isInline = false,
+} ) {
+	const TagName = isInline ? 'span' : 'div';
```
Comment (gziolo) [inline packages/components/src/visually-hidden/index.js]: Reakit uses an interesting pattern, they have `as` prop which you can pass to change the tag name, or even use a component. See: https://reakit.io/docs/composition/#as-prop

```js
<VisuallyHidden as="span">
	This should not show.
</VisuallyHidden>
```

They also offer an alternative way to use render props. See: https://reakit.io/docs/composition/#render-props

```js
< VisuallyHidden>
      { (  { children } ) => (
        <span {...props}>{ children }</span> 
      ) }
</VisuallyHidden>
```

@diegohaz, how much work would it be to replicate that here?

**Review by gziolo — CHANGES_REQUESTED** [review]
The missing CSS styles for the focused state is essential to make it work.

File: packages/components/src/visually-hidden/index.js (hunk 5 above)
Comment (youknowriad) [reply packages/components/src/visually-hidden/index.js]: `as` seem to "web" focused for me especially if we want the API of these components to be DOM agnostic. and the render props are great but I'd like to know if there's a potential use-case that the current API doesn't solve?

File: packages/components/src/screen-reader-text/index.js (hunk 3 above)
Comment (gziolo) [reply packages/components/src/screen-reader-text/index.js]: I think we have `label`, `p` and `leged` with the `screen-reader-text` class name as well.

We can iterate if you prefer to tackle it separately. See my related comment: https://github.com/WordPress/gutenberg/pull/18022#discussion_r337530787.

Comment (youknowriad) [reply packages/components/src/screen-reader-text/index.js]: Are all these solved by nesting VisuallyHidden inline inside the tags?

File: packages/components/src/visually-hidden/style.scss (hunk 4 above)
Comment (mkaz) [reply packages/components/src/visually-hidden/style.scss]: oh yeh, good catch

File: packages/components/src/screen-reader-text/index.js (hunk 3 above)
Comment (youknowriad) [reply packages/components/src/screen-reader-text/index.js]: I think we should decide now whether we want `isInline` prop or a render prop. I feel they address the same issue and I don't like API duplication :)

Comment (gziolo) [reply packages/components/src/screen-reader-text/index.js]: I don't think it solves the issue because the wrapping `label` or `legend` can have additional visual styles applied which you want to remove with `VisuallyHidden`. We would have to double-check that though. I'm making the case based on my assumptions :)

File: packages/components/src/visually-hidden/index.js (hunk 5 above)
Comment (gziolo) [reply packages/components/src/visually-hidden/index.js]: In Reakit `as` can be a string or Component, the native version:

```js
<VisuallyHidden as={ View }>
    This should not show.
</VisuallyHidden>
```

Comment (youknowriad) [reply packages/components/src/visually-hidden/index.js]: mmm interesting

File: packages/components/src/screen-reader-text/index.js (hunk 3 above)
Comment (mkaz) [reply packages/components/src/screen-reader-text/index.js]: You can look at `BaseControl` for an example usage. That solution inserts `<label className='screen-reader-text'>...</label>` 
This makes me think we would want to use the `as` prop. So the implementation would be `<VisuallyHidden as="label">...</VisuallyHidden>`

Also looks like a bug that if you pass in `hideLabelFromVision` prop but do not pass an `id` than the `BaseControl.VisualLabel` will be used, so would not include the class.

File: packages/components/src/visually-hidden/index.js (hunk 5 above)
Comment (diegohaz) [reply packages/components/src/visually-hidden/index.js]: It's simple to implement:
```jsx
function renderAsRenderProps({ as: T = "div", ...props }) {
  if (typeof props.children === "function") {
    return props.children(props);
  }
  return <T {...props} />;
}

function Component({ as = "span", ...props }) {
  return renderAsRenderProps({ as, ...props });
}
```

Using children as a function is better than the `as` prop to compose multiple components:
```jsx
<Component>
  {props => <div><Button {...props} /></div>}
</Component>
```

Comment (diegohaz) [reply packages/components/src/visually-hidden/index.js]: The `as` prop is quite popular nowadays. The first library I saw using it was [Semantic UI](https://react.semantic-ui.com/#augmentation). Not sure if there are other libraries combining it with render props though. But I think it's a simple and powerful combination.

Comment (mkaz) [reply packages/components/src/visually-hidden/index.js]: Thanks @diegohaz - I updated the component to use the "as" method as described, works well.

I defined the `renderAsRenderProps` function within the VisuallyHidden component, not exported. It  seems useful that we may want to use it elsewhere. If so, where would I move it to?

File: packages/components/src/visually-hidden/index.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+} ) {
+	return renderAsRenderProps( {
+		as,
+		className: 'components-visually-hidden',
+		...props,
+	} );
+}
+
+// render "as" to be either string (tag name) or component (fn)
+function renderAsRenderProps( { as: T = 'div', ...props } ) {
```
Comment (youknowriad) [inline packages/components/src/visually-hidden/index.js]: I like that pattern a lot and this can probably be extracted in a utils folder in the components package. That said a `<Component as="div" />` doesn't mean anything for mobile since there's no "div" on mobile and there's the expectation that the same props should be valid in all platform.

Not a blocker since mobile can decide to just ignore the "as"  prop but I thought I'd raise for awareness.

Comment (mkaz) [reply packages/components/src/visually-hidden/index.js]: I moved it to a `utils.js` file.

The div is just the default value, if used on Mobile they can pass in something different, or not use :-)

Comment (youknowriad) [reply packages/components/src/visually-hidden/index.js]: The idea is that mobile and web use the exact same code in the `edit` function of blocks... So there's no mobile do that and web do this.

Comment (mkaz) [reply packages/components/src/visually-hidden/index.js]: @youknowriad ah, I see.  Sounds like its not a blocker though, so should be fine.
Is the utils.js extraction fine?

Comment (mkaz) [reply packages/components/src/visually-hidden/index.js]: Actually @gziolo has the change requested for the CSS. Greg, when you get the chance, can you do a quick review again. I think everything has been addressed. Thanks! :bowing_man:

**Comment by diegohaz** [comment]
It looks great! I'd say only that the `renderAsRenderProps` util could be kept inside the component folder for now. There is a few things we should define before making it available for other components (naming, React Native version etc.), and that could be done in another PR. If this PR is merged and people start using the util in other components, it'll be harder to change later.

**Comment by mkaz** [comment]
@diegohaz Thanks. I moved the utils file, inside VisuallyHidden component folder for now. We can adjust as things get defined.

File: packages/components/src/visually-hidden/utils.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ * renderAsRenderProps is used to wrap a component and convert
+ * the passed property "as" either a string or component, to the
+ * rendered tag if a string, or component.
+ *
+ * See VisuallyHidden hidden for example.
+ *
+ * @param {string|Component} as what tag/component to render
+ * @return {Component} rendered component
+ */
+function renderAsRenderProps( { as: T = 'div', ...props } ) {
```
Comment (gziolo) [inline packages/components/src/visually-hidden/utils.js]: Can we avoid using shortcuts, please. It makes it hard to follow for those not familiar with the pattern:
```suggestion
function renderAsRenderProps( { as: Component = 'div', ...props } ) {
```

File: packages/components/src/visually-hidden/utils.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+/**
+ * renderAsRenderProps is used to wrap a component and convert
+ * the passed property "as" either a string or component, to the
+ * rendered tag if a string, or component.
+ *
+ * See VisuallyHidden hidden for example.
+ *
+ * @param {string|Component} as what tag/component to render
```
Comment (gziolo) [inline packages/components/src/visually-hidden/utils.js]: Nit:
```suggestion
 * @param {string|Component} as A tag or component to render.
```

File: packages/components/src/visually-hidden/utils.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ * @return {Component} rendered component
```
Comment (gziolo) [inline packages/components/src/visually-hidden/utils.js]:
```suggestion
 * @return {Component} The rendered component.
```

File: packages/components/src/visually-hidden/index.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (gziolo) [inline packages/components/src/visually-hidden/index.js]: This empty space is not necessary. Why is ESLint not complaining? :) We really need Prettier 😍

File: packages/components/src/visually-hidden/style.scss [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+	margin: -1px;
+	overflow: hidden;
+	padding: 0;
+	position: absolute;
+	width: 1px;
+	word-wrap: normal !important;
+}
+
+.components-visually-hidden:focus {
+	background-color: #eee;
```
Comment (gziolo) [inline packages/components/src/visually-hidden/style.scss]: Can this be converted to one of the variables used in SCSS files?

File: packages/components/src/visually-hidden/style.scss [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+	display: block;
+	font-size: 1em;
+	height: auto;
+	left: 5px;
+	line-height: normal;
+	padding: 15px 23px 14px;
+	text-decoration: none;
+	top: 5px;
+	width: auto;
+	z-index: 100000;
```
Comment (gziolo) [inline packages/components/src/visually-hidden/style.scss]: In the context of Gutenberg, we probably should use a special helper which processes all `z-indexes`:

https://github.com/WordPress/gutenberg/blob/7515397fef469e8cbdaf68162d6d6e1f2a5f9aa1/packages/base-styles/_z-index.scss#L127-L133

**Review by gziolo — APPROVED** [review]
I listed a few nits to address, those aren't blockers if you plan to work on a follow-up PR. Your call @mkaz.

Thank you for addressing all the feedback, it was a journey to get us to this stage but I think it's worth it. Props to @diegohaz for help with the `renderAsRenderProps` integration 💯

File: packages/components/src/visually-hidden/style.scss (hunk 11 above)
Comment (mkaz) [reply packages/components/src/visually-hidden/style.scss]: This raises another issue, it doesn't look like the SCSS variables are defined within Components package, or if they are I've missed where. It seems they rely on Block Editor styling, which can be used, but not sure where this separation should be. So this might need to become a separate issue for how to define and use SCSS with Components.

File: packages/components/src/visually-hidden/index.js (hunk 10 above)
Comment (mkaz) [reply packages/components/src/visually-hidden/index.js]: Funny because the space should be below, newline at the end.

File: packages/components/src/visually-hidden/style.scss (hunk 11 above)
Comment (youknowriad) [reply packages/components/src/visually-hidden/style.scss]: That's an already solved problem. variables are now defined in the `base-styles` package and these are made available to all packages using SASS including the @wordpress/components package.

Comment (mkaz) [reply packages/components/src/visually-hidden/style.scss]: :+1: Great. So sounds like mostly a matter of documentation and examples for the case that someone wants to use the components package, show they would define their own or use base-styles package.
